compiler/rustc_driver_impl/src/args.rs

```diff
@@ -1,52 +1,62 @@
+use rustc_session::config::ErrorOutputType;
+use rustc_span::ErrorGuaranteed;
 use std::error;
 use std::fmt;
 use std::fs;
 use std::io;
 
-fn arg_expand(arg: String) -> Result<Vec<String>, Error> {
+fn arg_expand(arg: &str) -> Result<Vec<String>, Error> {
     if let Some(path) = arg.strip_prefix('@') {
         let file = match fs::read_to_string(path) {
             Ok(file) => file,
             Err(ref err) if err.kind() == io::ErrorKind::InvalidData => {
-                return Err(Error::Utf8Error(Some(path.to_string())));
+                return Err(Error::Utf8Error(path.to_string()));
             }
             Err(err) => return Err(Error::IOError(path.to_string(), err)),
         };
         Ok(file.lines().map(ToString::to_string).collect())
     } else {
-        Ok(vec![arg])
+        Ok(vec![arg.to_string()])
     }
 }
 
+/// Replaces any `@file` arguments with the contents of `file`, with each line of `file` as a
+/// separate argument.
+///
 /// **Note:** This function doesn't interpret argument 0 in any special way.
 /// If this function is intended to be used with command line arguments,
 /// `argv[0]` must be removed prior to calling it manually.
-pub fn arg_expand_all(at_args: &[String]) -> Vec<String> {
-    let mut args = Vec::new();
+pub fn arg_expand_all(at_args: &[String]) -> Result<Vec<String>, ErrorGuaranteed> {
+    let mut res = Ok(Vec::new());
     for arg in at_args {
-        match arg_expand(arg.clone()) {
-            Ok(arg) => args.extend(arg),
-            Err(err) => rustc_session::early_error(
-                rustc_session::config::ErrorOutputType::default(),
-                format!("Failed to load argument file: {err}"),
-            ),
+        match arg_expand(arg) {
+            Ok(arg) => {
+                if let Ok(args) = &mut res {
+                    args.extend(arg)
+                }
+            }
+            Err(err) => {
+                res = Err(rustc_session::early_error_no_abort(
```

Comment:
I don't understand the purpose.
I'd understand, if we reported this error and then recovered and continued collecting remaining arguments, but here the error is immediately bubbled up so it's not clear what is the difference from the old behavior.

Reply:
Same issue with invalid Unicode errors in another commit.